projects/dataset_recorder/robot_recorder.h

```diff
@@ -0,0 +1,225 @@
+// Standard C++ includes
+#include <array>
+#include <chrono>
+#include <exception>
+#include <memory>
+#include <thread>
+
+namespace BoBRobotics {
+class RobotRecorder {
+public:
+    RobotRecorder(GPS::GPSReader &gps, bool useSystemClock)
+      : m_GPS{ gps }
+      , m_Camera{ getCamera() }
+      , m_Database{ getCurrentDateTime(useSystemClock) }
+      , m_Recorder{ m_Database.getRouteVideoRecorder(m_Camera->getOutputSize(),
+                                                     getFrameRate(),
+                                                     "mp4",
+                                                     "mp4v",
+                                                     { "Pitch [degrees]",
+                                                       "Roll [degrees]",
+                                                       "Speed",
+                                                       "Steering angle [degrees]",
+                                                       "UTM zone",
+                                                       "GPS quality",
+                                                       "Horizontal dilution [mm]",
+                                                       "Timestamp [ms]" }) }
+    {
+        // Start timing
+        m_Stopwatch.start();
+    }
+
+    ~RobotRecorder()
+    {
+        // Make sure that written data is actually written to disk before we exit
+        ::sync();
+    }
+
+    //! Record from sensors. Returns time elapsed since start of recording.
+    template<class T>
+    auto step(T &robot, BN055 &imu)
+    {
+        using namespace units::angle;
+        using namespace units::length;
+        using namespace units::time;
+
+        // Read image from camera (synchronously)
+        m_Camera->readFrameSync(m_Frame);
+
+#ifdef DUMMY_CAMERA
+        // RandomInputs don't block, so throttle the framerate to 30 fps
+        std::this_thread::sleep_for(std::chrono::milliseconds(33));
+#endif
+
+        // Sync time to when camera frame was read
+        const millisecond_t time = m_Stopwatch.elapsed();
+
+        // get imu data
+        degree_t yaw{ NAN }, pitch{ NAN }, roll{ NAN };
+        try {
+            const auto angles = imu.getEulerAngles();
+            yaw = angles[0];
+            pitch = angles[1];
+            roll = angles[2];
+        } catch (std::exception &e) {
+            LOGE << "Could not read from IMU: " << e.what();
+        }
+
+        // read speed and steering angle from robot
+        float botSpeed = NAN;
+        degree_t turnAngle{ NAN };
+        try {
+            std::tie(botSpeed, turnAngle) = robot.readRemoteControl();
+        } catch (std::exception &e) {
+            // if we can't read speed or angle, we just write nan values
+            LOGE << "Could not read speed and steering angle from robot : " << e.what();
+        }
+
+        // Poll for GPS data
+        if (const auto optData = m_GPS.read()) {
+            const auto &gpsData = optData.value();
```

Comment:
should this be ``m_GPSData = optData.value()``? If not, I don't understand where ``m_GPSData`` is getting set

Reply:
Oooops! Good spot

Reply:
Yey! I actually spotted a bug!